### PR TITLE
Ports Pixel Projectiles from Aurora.3

### PR DESCRIPTION
**Pull request by Mazianni**
Ports Aurora's pixel projectile system.

Adds SSProjectiles and all the fancy math related to the projectile system, too. 

Do you know what you're doing? Do you want to help? _Thank god because I BARELY know what I'm doing here._

### CODE REVIEW

File: code/_helpers/maths.dm [hunk 1]
```diff
@@ -134,3 +134,6 @@
 
 /matrix/proc/get_angle()
 	return Atan2(b,a)
+
+/proc/Modulus(x, y)
```
Comment (MistakeNot4892) [inline code/_helpers/maths.dm]: Could do this as
````
#define MODULUS_FLOAT(X, Y) ( (X) - (Y) * round((X) / (Y)) )
````
as an optimization/to make it more clear why it's a thing when `%` exists.

**Comment by Mazianni** [comment]
Status:
It compiles now! Good.

The downside:
Bullet-based guns don't work. They do all the fancy firing stuff but dont' actually ... throw a bullet.
Lasers work! ... But they only fire north.
And you can't shoot yourself.

**Comment by Mazianni** [comment]
Handled the issue with guns not working quite right.

Still need to fix being able to shoot yourself, and re-implement accuracy (which wasn't a thing pixel projectiles supported?)

Bullets and tracers look weird.

**Comment by Mazianni** [comment]
Tracers more or less fixed. Guns will need some dispersion tweaks with the new system. Bullets still look funky for whatever reason.

File: code/game/objects/effects/projectile/projectile_effects.dm [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/obj/effect/projectile
+	name = "pew"
+	icon = 'icons/obj/projectiles.dmi'
+	icon_state = "nothing"
+	layer = 4.5
```
Comment (comma) [inline code/game/objects/effects/projectile/projectile_effects.dm]: what's with the magical layer?

File: code/controllers/subsystems/processing/projectiles.dm [hunk 3]
```diff
@@ -0,0 +1,6 @@
+PROCESSING_SUBSYSTEM_DEF(projectiles)
+	name = "Projectiles"
+	priority = SS_PRIORITY_PROJECTILES
+	flags = SS_TICKER|SS_NO_INIT
```
Comment (Lohikar) [inline code/controllers/subsystems/processing/projectiles.dm]: Pls test this as a non-SS_TICKER SS with 1 wait, I'd like to keep tickers to a minimum.

File: code/game/objects/effects/projectile/projectile_tracer.dm [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+/obj/effect/projectile/tracer/emitter
+	icon_state = "emitter"
+	light_color = LIGHT_COLOR_GREEN
+
+/obj/effect/projectile/tracer/tachyon
+	name = "particle beam"
+	icon_state = "invisible"
+	light_color = LIGHT_COLOR_VIOLET
+
+/obj/effect/projectile/tracer/bfg
```
Comment (Lohikar) [inline code/game/objects/effects/projectile/projectile_tracer.dm:84]: Do we even have this?

File: code/modules/projectiles/projectile.dm [hunk 5]
```diff
@@ -66,11 +65,30 @@
...
 	var/vacuum_traversal = 1 //Determines if the projectile can exist in vacuum, if false, the projectile will be deleted if it enters vacuum.
-
-	var/datum/plot_vector/trajectory	// used to plot the path of the projectile
-	var/datum/vector_loc/location		// current location of the projectile in pixel space
-	var/matrix/effect_transform			// matrix to rotate and scale projectile effects - putting it here so it doesn't
-										//  have to be recreated multiple times
+//Movement parameters
+	var/speed = 0.4		//Amount of deciseconds it takes for projectile to travel
+	var/pixel_speed = 33	//pixels per move - DO NOT FUCK WITH THIS UNLESS YOU ABSOLUTELY KNOW WHAT YOU ARE DOING OR UNEXPECTED THINGS /WILL/ HAPPEN!
+	var/Angle = 0
```
Comment (Lohikar) [inline code/modules/projectiles/projectile.dm:71]: Lower-case.

File: code/game/objects/effects/projectile/projectile_effects.dm [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+/obj/effect/projectile/singularity_pull()
+	return
+
+/obj/effect/projectile/singularity_act()
+	return
+
+/obj/effect/projectile/proc/scale_to(nx,ny,override=TRUE)
+	var/matrix/M
+	if(!override)
```
Comment (Lohikar) [inline code/game/objects/effects/projectile/projectile_effects.dm]: Nit: flip the arms of this if to remove the not.

File: code/game/objects/effects/projectile/projectile_effects.dm [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		M = transform
+	else
+		M = new
+	M.Scale(nx,ny)
+	transform = M
+
+/obj/effect/projectile/proc/turn_to(angle,override=TRUE)
+	var/matrix/M
+	if(!override)
```
Comment (Lohikar) [inline code/game/objects/effects/projectile/projectile_effects.dm]: Nit: same as above.

File: code/modules/projectiles/projectile.dm [hunk 8]
```diff
@@ -524,9 +349,299 @@
...
+		qdel(src)
+	if(angle_offset)
+		setAngle(Angle + angle_offset)
+
+/obj/item/projectile/Crossed(atom/movable/AM) //A mob moving on a tile with a projectile is hit by it.
+	..()
+	if(isliving(AM) && (AM.density || AM == original) && !(pass_flags & PASS_FLAG_MOB))
+		Bump(AM)
+
+/obj/item/projectile/Initialize()
```
Comment (Lohikar) [inline code/modules/projectiles/projectile.dm]: Just define `permutated` as `= list()`, having a proc just to initialize it is pointless.

File: code/modules/projectiles/projectile.dm [hunk 9]
```diff
@@ -524,9 +349,299 @@
...
+		var/list/screenview = getviewsize(user.client.view)
+		var/screenviewX = screenview[1] * world.icon_size
+		var/screenviewY = screenview[2] * world.icon_size
+
+		var/ox = round(screenviewX/2) - user.client.pixel_x //"origin" x
+		var/oy = round(screenviewY/2) - user.client.pixel_y //"origin" y
+		angle = Atan2(y - oy, x - ox)
+	return list(angle, p_x, p_y)
+
+/obj/item/projectile/proc/Range()
```
Comment (Lohikar) [inline code/modules/projectiles/projectile.dm]: This proc could be better named.

File: code/modules/projectiles/projectile.dm [hunk 10]
```diff
@@ -524,9 +349,299 @@
...
+
+/obj/item/projectile/proc/process_hitscan()
+	var/safety = range * 3
+	var/return_vector = RETURN_POINT_VECTOR_INCREMENT(src, Angle, MUZZLE_EFFECT_PIXEL_INCREMENT, 1)
+	record_hitscan_start(return_vector)
+	while(loc && !QDELETED(src))
+		if(paused)
+			stoplag(1)
+			continue
+		if(safety-- <= 0)
```
Comment (Lohikar) [inline code/modules/projectiles/projectile.dm]: No `--`s in flow control please, pull the `safety` decrement into its own line above.

File: code/modules/projectiles/projectile.dm [hunk 11]
```diff
@@ -524,9 +349,299 @@
...
+			qdel(src)
+			crash_with("WARNING: [type] projectile encountered infinite recursion during hitscanning in [__FILE__]/[__LINE__]!")
```
Comment (Lohikar) [inline code/modules/projectiles/projectile.dm]: I'm pretty sure that crash_with already includes file/line.

File: code/modules/projectiles/projectile/trace.dm [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+		return 0
+
+	var/obj/item/projectile/test/trace = new /obj/item/projectile/test(get_turf(firer)) //Making the test....
+
+	//Set the flags and pass flags to that of the real projectile...
+	trace.pass_flags = pass_flags
+
+	return trace.launch(target) //Test it!
+
+/obj/item/projectile/proc/_check_fire(atom/target as mob, var/mob/living/user as mob)  //Checks if you can hit them or not.
```
Comment (Lohikar) [inline code/modules/projectiles/projectile/trace.dm]: What's this for? I don't see any references to it.

**Comment by MistakeNot4892** [comment]
Gentle poke, please update the PR.

File: code/game/objects/effects/projectile/projectile_effects.dm (hunk 2 above)
Comment (Mazianni) [reply code/game/objects/effects/projectile/projectile_effects.dm]: I have no idea what this layer is for, but I assume it's important.

File: code/controllers/subsystems/processing/projectiles.dm (hunk 3 above)
Comment (Mazianni) [reply code/controllers/subsystems/processing/projectiles.dm]: I am not entirely sure what you're asking me to do here.

File: code/game/objects/effects/projectile/projectile_tracer.dm (hunk 4 above)
Comment (Mazianni) [reply code/game/objects/effects/projectile/projectile_tracer.dm:84]: Probably not.

File: code/modules/projectiles/projectile/trace.dm (hunk 12 above)
Comment (Mazianni) [reply code/modules/projectiles/projectile/trace.dm]: I believe it's used for raycasting projectiles, I think?

File: code/modules/projectiles/projectile.dm (hunk 5 above)
Comment (Mazianni) [reply code/modules/projectiles/projectile.dm:71]: In next push.

File: code/modules/projectiles/projectile.dm (hunk 8 above)
Comment (Mazianni) [reply code/modules/projectiles/projectile.dm]: It was already defined as a list, oddly enough. This was a direct port, so assume auroracode oddities.

File: code/modules/projectiles/projectile.dm (hunk 9 above)
Comment (Mazianni) [reply code/modules/projectiles/projectile.dm]: Renamed in next push.

File: code/modules/projectiles/projectile.dm (hunk 10 above)
Comment (Mazianni) [reply code/modules/projectiles/projectile.dm]: Done.

File: code/modules/projectiles/projectile/trace.dm (hunk 12 above)
Comment (Mazianni) [reply code/modules/projectiles/projectile/trace.dm]: Correction, this does nothing. Removing.

File: code/game/objects/effects/projectile/projectile_effects.dm (hunk 7 above)
Comment (Mazianni) [reply code/game/objects/effects/projectile/projectile_effects.dm]: What do you mean by this?

File: code/game/objects/effects/projectile/projectile_effects.dm (hunk 6 above)
Comment (Mazianni) [reply code/game/objects/effects/projectile/projectile_effects.dm]: What do you mean by this?

File: code/__defines/flags.dm [hunk 13]
```diff
@@ -55,9 +55,11 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 #define PASS_FLAG_TABLE                   0x1
 #define PASS_FLAG_GLASS                   0x2
 #define PASS_FLAG_GRILLE                  0x4
+#define PASS_FLAG_MOB					  0x8
```
Comment (MistakeNot4892) [inline code/__defines/flags.dm]: Indent with spaces please, it's out of line with the others. :(

File: code/datums/position_point_vector.dm [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+/datum/point/proc/initialize_location(tile_x, tile_y, tile_z, p_x = 0, p_y = 0)
+	if(!isnull(tile_x))
+		x = ((tile_x - 1) * world.icon_size) + world.icon_size / 2 + p_x + 1
+	if(!isnull(tile_y))
+		y = ((tile_y - 1) * world.icon_size) + world.icon_size / 2 + p_y + 1
+	if(!isnull(tile_z))
+		z = tile_z
+
+/datum/point/proc/debug_out()
```
Comment (MistakeNot4892) [inline code/datums/position_point_vector.dm]: This proc doesn't seem to be used anywhere.

File: code/datums/position_point_vector.dm [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+/datum/point/proc/return_turf()
+	return locate(Ceiling(x / world.icon_size, 1), Ceiling(y / world.icon_size, 1), z)
+
+/datum/point/proc/return_coordinates()		//[turf_x, turf_y, z]
+	return list(Ceiling(x / world.icon_size, 1), Ceiling(y / world.icon_size, 1), z)
+
+/datum/point/proc/return_position()
+	return new /datum/position(src)
+
+/datum/point/proc/return_px()
```
Comment (MistakeNot4892) [inline code/datums/position_point_vector.dm:122]: Not sure if I am missing something but `return_px` and `return_py` could both be #defines to go alongside `MODULUS_FLOAT`, as-is there is no performance benefit to using a define for MODULUS_FLOAT as these procs have overhead.

**Review by MistakeNot4892 — COMMENTED** [review]
Generally this looks alright to me, but I only skimmed most of the mathematics as I don't know much about trig or trajectories or such.

**Comment by MistakeNot4892** [comment]
image
`:thinking emoji:`

**Comment by MistakeNot4892** [comment]
This still needs a fix applied for passing colour to the muzzle flash and impact decals (used by capacitor rifles) and possibly something done about the weird only-put-light-in-the-middle aspect.

**Comment by MistakeNot4892** [comment]
Poking it to reset actions.

**Comment by MistakeNot4892** [comment]
Moved to a new PR as discussed on Discord.

**Pull request closed without merging**